feat(react-native): refactor docs about native permissions (#865)
GetStream/stream-video-js: packages/react-native-sdk/docusaurus/docs/reactnative/03-core/08-native-permissions.mdx

title: Manage Native Permissions
---

> Rendering a video call or audio call requires native platform permissions, which is mandatory to access a camera and microphone respectively. The SDK needs to know to the state of camera and microphone permissions to get video and audio streams. We can inform the states to the SDK through the following:

Reviewer: ```suggestion
Rendering a video call or audio call requires native platform permissions, which are mandatory to access a camera and microphone respectively. The SDK needs to know the state of the permission of the camera and microphone to get video and audio streams. We can inform the states to the SDK through the following:
```

```ts
import { StreamVideoRN } from '@stream-io/video-react-native-sdk';

StreamVideoRN.setPermissions({
  isCameraPermissionGranted: true,
  isMicPermissionGranted: true,
});
```

In this guide, we will create a hook named `useSyncPermissions` to manage the native permissions required for the app. This hook will be used to request the app's permissions and inform them to the SDK.

> The hook will request for the relevant permissions when the screen is mounted and also informs the permission state to the SDK when the app state changes to the foreground. This ensures the latest permissions are synced with the SDK, that way the SDK can get video and audio streams for a call.

Reviewer: ```suggestion
The hook will request the relevant permissions when the screen is mounted and also informs the permission state to the SDK when the app state changes to the foreground. This ensures the latest permissions are synced with the SDK. That way the SDK can get video and audio streams for a call.
```

Additionally, to request permissions in both platforms easily we will use the [`react-native-permissions`](https://github.com/zoontek/react-native-permissions) library. You can run the following command to install it:

```bash title=Terminal
yarn add react-native-permissions
```

#### iOS

Additionally, for iOS, you need update your `package.json` by adding the permissions used in your app.

```js title=package.json
{
  "reactNativePermissionsIOS": [
    "Camera",
    "Microphone",
  ],
}
```

After that, we need to run the following commands:

```bash title=Terminal
npx react-native setup-ios-permissions
npx pod-install
```

## Step 1 - Add functions to update permissions to the SDK

In this step, we create two functions called `androidProcessPermissions` and `iOSProcessPermissions`. These functions are responsible for informing the status of the permission grants to the SDK. 

```ts title=src/utils/updatePermissions.ts
import { PERMISSIONS, RESULTS, PermissionStatus } from 'react-native-permissions';
import { StreamVideoRN } from '@stream-io/video-react-native-sdk';

// The extracted type of the return value of `requestMultiple`

> type Results = Awaited<ReturnType<typeof requestMultiple>>;

Reviewer: Was trying the tutorial locally, but this seems to be unused in the updatePermissions.ts.

Can you check?

Author: yes, forgot to remove this line

    const results = await checkMultiple([
      PERMISSIONS.ANDROID.CAMERA,
      PERMISSIONS.ANDROID.RECORD_AUDIO,
    ]);
    // Sync the permissions with the Stream Video SDK
    androidProcessPermissions(results);
  }
};
```

## Step 4 - Create the hook to manage permissions

In the step, we create the `useSyncPermissions` hook. As mentioned earlier, this hook will be used to request the app's permissions and inform them to the SDK. 

Additional to updating permissions after requesting them, it is also necessary to check the status of permissions when app comes to foreground. This will ensure that the permissions are updated when the user comes back to the app from the background after changing the permissions in the native settings screen. In order to do this, we will also add a listener to the app state change event in this hook.

```tsx
import { useEffect, useRef } from 'react';
import { AppState } from 'react-native';
import { requestAndUpdatePermissions } from 'src/utils/requestAndUpdatePermissions';
import { checkAndUpdatePermissions } from 'src/utils/checkAndUpdatePermissions';

export const useSyncPermissions = () => {
  // request permissions on mount

>   useEffect(() => {
>     requestAndUpdatePermissions();
>   }, []);

Reviewer: ```suggestion
   useEffect(() => {
    const requestAndUpdate = async () => {
      await requestAndUpdatePermissions();
    };
    requestAndUpdate();
  }, []);
```

Since `requestAndUpdatePermissions` is asynchronous. Feel free to use an appropriate name here.

Author: both snippets do the same, nothing is different

  // check permissions on foreground
  const appStateRef = useRef(AppState.currentState);
  useEffect(() => {
    const subscription = AppState.addEventListener('change', nextAppState => {
      if (

>         appState.current.match(/inactive|background/) &&

Reviewer: ```suggestion
        appStateRef.current.match(/inactive|background/) &&
```

        nextAppState === 'active'
      ) {
        checkAndUpdatePermissions();;
      }

>       appState.current = nextAppState;

Reviewer: ```suggestion
      appStateRef.current = nextAppState;
```

    });

    return () => {
      subscription.remove();
    };
  }, []);
};
```

## Step 5 - Use the Hook

> In this final step, we use the `useSyncPermissions` hook in the screen of our choice. As an example below, we use it in the screen where we pass the `call` object to the SDK.

Author: No need to use at the earliest, we use rxjs internally, to listen to permissions and then if granted to enable audio stream for example
